Signal/src/network/GiphyDownloader.swift

```diff
@@ -227,93 +434,220 @@ extension URLSessionTask {
         }
     }
 
-    private func startRequestIfNecessary() {
+    private func removeAssetRequestFromQueue(assetRequest: GiphyAssetRequest) {
+        AssertIsOnMainThread()
+
+        guard assetRequestQueue.contains(assetRequest) else {
+            Logger.warn("\(TAG) could not remove asset request from queue: \(assetRequest.rendition.url)")
+            return
+        }
+
+        assetRequestQueue = assetRequestQueue.filter { $0 != assetRequest }
+    }
+
+    // Start a request if necessary, complete asset requests if possible.
+    private func processRequestQueue() {
         AssertIsOnMainThread()
 
         DispatchQueue.main.async {
-            guard self.activeAssetRequests.count < self.kMaxAssetRequestCount else {
-                return
-            }
             guard let assetRequest = self.popNextAssetRequest() else {
                 return
             }
             guard !assetRequest.wasCancelled else {
                 // Discard the cancelled asset request and try again.
-                self.startRequestIfNecessary()
+                self.processRequestQueue()
                 return
             }
             guard UIApplication.shared.applicationState == .active else {
                 // If app is not active, fail the asset request.
+                assetRequest.state = .failed
                 self.assetRequestDidFail(assetRequest:assetRequest)
-                self.startRequestIfNecessary()
+                self.processRequestQueue()
                 return
             }
 
-            self.activeAssetRequests.insert(assetRequest)
-
             if let asset = self.assetMap.get(key:assetRequest.rendition.url) {
                 // Deferred cache hit, avoids re-downloading assets that were
                 // downloaded while this request was queued.
 
+                assetRequest.state = .complete
                 self.assetRequestDidSucceed(assetRequest : assetRequest, asset: asset)
                 return
             }
 
             guard let downloadSession = self.giphyDownloadSession() else {
                 owsFail("\(self.TAG) Couldn't create session manager.")
+                assetRequest.state = .failed
                 self.assetRequestDidFail(assetRequest:assetRequest)
                 return
             }
 
+            if assetRequest.state == .waiting {
+                // If asset request hasn't yet determined the resource size,
+                // try to do so now.
+                assetRequest.state = .requestingSize
+
+                var request = URLRequest(url: assetRequest.rendition.url as URL)
+                request.httpMethod = "HEAD"
+
+                let task = downloadSession.dataTask(with:request, completionHandler: { [weak self] _, response, error -> Void in
+                    self?.handleAssetSizeResponse(assetRequest:assetRequest, response:response, error:error)
+                })
+
+                task.resume()
+                return
```

Comment:
Do we need to kick the run loop here?

Else won't we potentially leave the queue hanging or less-than-full?

And same question for the end of this method?

Reply:
Already fixed!